### PR TITLE
[ADD][12.0] sale_invoice_plan

### PR DESCRIPTION
**Pull request by kittiu**
By standard feature, user can gradually create partial invoices, one by one.
This module add ability to create invoices based on the predefined invoice plan,
either all at once, or one by one.
The plan support both advance invoice and installment invoices.

image

### CODE REVIEW

**Comment by pedrobaeza** [comment]
This should be checked by reviewers before merging.

**Comment by kittiu** [comment]
Any comments for this module?

File: sale_invoice_plan/models/sale.py
Comment (nikul-serpentcs) [inline sale_invoice_plan/models/sale.py]: @kittiu IMO I think here change field name it looks like a function

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Improve Code

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py]: @nikul-serpentcs thanks for your comment.

**Comment by kittiu** [comment]
This module has been here a while. Any more comments?

**Review by jgrandguillaume — APPROVED** [review]
Functional test: Ok, thanks for the contribution.
:+1:

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [inline sale_invoice_plan/models/sale.py]: You should use ValidationError in constrains

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [inline sale_invoice_plan/models/sale.py]: You could use this :

```suggestion
        if self.filtered(lambda r: r.use_invoice_plan and not r.invoice_plan_ids):
            raise...
```

just to avoid indent

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [inline sale_invoice_plan/models/sale.py:66]: This could be set in an helper computed field no ?

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [inline sale_invoice_plan/models/sale.py]: It's better to use:

```suggestion
            from odoo.fields import first
            order_line = first(line.sale_line_ids)
```

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [inline sale_invoice_plan/models/sale.py:216]: Is the plan_qty could be set in an helper field with digit precision as the following comparison should be done with float_compare?

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [inline sale_invoice_plan/models/sale.py:115]: IMHO in '_compute_new_invoice_quantity', you should do a write on invoice values.

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. Some comments.

Looks great!

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py:115]: Sorry, I don't quite understand. You mean moving 
`plan.invoice_ids += invoice`
inside _compute_new_invoice_quantity()?

Comment (rousseldenis) [reply sale_invoice_plan/models/sale.py:115]: No, maybe I'm at the wrong place to do the comment.

But here you are in an action (launched by a button I suppose).

that's why you have to use write() in '_compute_new_invoice_quantity' instead of doing 'line.quantity = plan_qty'

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py:216]: @rousseldenis sorry, I don't understand what you mean about -> **plan_qty could be set in an helper field with digit precision**

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py]: Done

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py:66]: @rousseldenis sorry, I don't understand what you mean.

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py]: Done

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py]: Done.

**Comment by AaronHForgeFlow** [comment]
It works fine. It would be great to include another option that allows to invoice every X days (X is entered by the user).

It also would be awesome to do partial deliveries according to a plan. But that would go to another module. Thank you @kittiu

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py:115]: Done, I changed to write()
By the way, I also curious. In what situation we should not use record.field = value?
Thank you

**Comment by kittiu** [comment]
> It works fine. It would be great to include another option that allows to invoice every X days (X is entered by the user).

@aheficent thanks for your comment. Currently, we can set X days (month, year) as interval. You are looking for something different?
image

**Comment by kittiu** [comment]
> It also would be awesome to do partial deliveries according to a plan. But that would go to another module. Thank you @kittiu

Yes, that might be sale_delivery_plan ? :)

**Comment by AaronHForgeFlow** [comment]
> sale_delivery_plan

Yes :)

**Comment by AaronHForgeFlow** [comment]
> > It works fine. It would be great to include another option that allows to invoice every X days (X is entered by the user).
> 
> @aheficent thanks for your comment. Currently, we can set X days (month, year) as interval. You are looking for something different?
> image

That's good, did not notice of that on runbot. Thanks.

**Review by AaronHForgeFlow — APPROVED** [review]
:+1:

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [reply sale_invoice_plan/models/sale.py:115]: Usually, such affectation should only be used in compute and inverse functions as the write is done after (with recomputes and so on). When it is launched through an action or,... you are not sure that a write will be done(and good recomputes on fields that depends on the one you've modified).

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [reply sale_invoice_plan/models/sale.py:66]: It was a minor remark but generally it is better to store some intermediate computed values in a field to lower the function code and to make it more readable.

File: sale_invoice_plan/models/sale.py
Comment (rousseldenis) [reply sale_invoice_plan/models/sale.py:216]: The field 'line.quantity' is a field with decimal precision. You must at least compare with 'float_compare'.

The remark about 'helper' field on 'plan_qty' is quite the same I did here above.

It's sometimes useful to make a computed field with a part of a function to make it more readable and sometimes reuse the field elsewhere.

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py:66]: Thank you @rousseldenis , I think I understand your point now :)
My design decision so far is to minimize the new field in sale.order (I only have 1 new column which is `use_invoice_plan`).
From what I see from this function, the useful compute field would be `percent` and `percent_last`. But to compute it, I need to store `num_installment` permanently in sale.order. I think it might not worth it yet? In the future, if there is the need of `num_installment` then it will be sound. WDYT?

File: sale_invoice_plan/models/sale.py
Comment (kittiu) [reply sale_invoice_plan/models/sale.py:216]: Done use float_compare(), thank you.
For the plan_qty as helper, as I try to look at, may be not very easy to do at this point.

**Review by rousseldenis — APPROVED** [review]
Code review

**Comment by AaronHForgeFlow** [comment]
Hi @nikul-serpentcs  IMHO your comments have been attended here, can you update your review please?

**Review by nikul-serpentcs — APPROVED** [review]
Code Review LGTM :+1:

**Comment by mohamedhagag** [comment]
just did a functional test and working fine, should we merge ?

**Comment by kittiu** [comment]
+1 :)

**Comment by rousseldenis** [comment]
@OCA/core-maintainers

**Comment by kittiu** [comment]
Hi @OCA/core-maintainers, this addon has been used and been pending here quite sometime. Would appreciate the merge :)

**Comment by mohamedhagag** [comment]
SWM ? :)

**Pull request closed without merging**
